bin/coqtop.ml

```diff
@@ -0,0 +1,92 @@
+open Stdune
+open Import
+
+let doc =
+  "Execute the Coq toplevel with the local configuration."
+
+let man =
+  [ `S "DESCRIPTION"
+  ; `P
+      {|$(b,dune coq top FILE -- ARGS) runs the Coq toplevel to process the
+          given $(b,FILE). The given arguments are completed according to the
+          local configuration. This is equivalent to running $(b,coqtop ARGS)
+          with a $(b,_CoqProject) file containing the local configurations
+          from the $(b,dune) files, but does not require maintaining a
+          $(b,_CoqProject) file.|}
+  ; `Blocks Common.help_secs
+  ]
+
+let info = Term.info "top" ~doc ~man
+
+let term =
+  let+ common = Common.term
+  and+ context =
+    Common.context_arg ~doc:{|Run the Coq toplevel in this build context.|}
+  and+ coq_file_arg =
+    Arg.(required & pos 0 (some string) None (Arg.info [] ~docv:"COQFILE"))
+  and+ extra_args =
+    Arg.(value & pos_right 0 string [] (Arg.info [] ~docv:"ARGS"))
+  in
+  let config = Common.init common in
+  (* Checks on the argument file. *)
+  let source = Common.prefix_target common coq_file_arg in
+  if not (Path.is_file (Path.of_string source)) then
+    User_error.raise [ Pp.textf "cannot find file: %s" coq_file_arg ];
+  let coqtop, argv, env =
+    Scheduler.go ~common ~config (fun () ->
+        let open Fiber.O in
+        let* setup = Import.Main.setup () in
+        let* setup = Memo.Build.run setup in
+        let sctx = Import.Main.find_scontext_exn setup ~name:context in
+        let context = Dune_rules.Super_context.context sctx in
+        let coq_file_build =
+          let p = Common.prefix_target common coq_file_arg in
+          Path.Build.relative context.build_dir p
+        in
+        let dir =
+          let dir = Filename.dirname coq_file_arg in
+          let p = Common.prefix_target common dir in
+          Path.Build.relative context.build_dir p
+        in
+        let* (coqtop, args) =
+          let open Memo.Build.O in
+          Build_system.run_exn (fun () ->
+              let* dc = Dune_rules.Dir_contents.get sctx ~dir in
+              let* coq_src = Dune_rules.Dir_contents.coq dc in
+              let coq_module =
+                let source = coq_file_build in
+                match Dune_rules.Coq_sources.find_module ~source coq_src with
+                | Some m -> m
+                | None -> User_error.raise
+                            [ Pp.textf "cannot find file: %s" coq_file_arg ]
+              in
+              let* _ : unit * Dep.Fact.t Dep.Map.t =
+                let deps =
+                  Dune_rules.Coq_rules.deps_of ~dir ~boot_type:No_boot coq_module
+                in
+                Action_builder.run deps Eager
+              in
+              let args =
+                [] (* FIXME arguments used when invoking coqc on the file. *)
+              in
+              let prog = "coqtop" in
```

Comment:
It would be nice if this was configurable, eg `dune coq idetop` would be a way to run `coqidetop` which is what vscode needs (or coqide, or coqtail...)

Reply:
indeed a --coq-command-name option could be great

Reply:
That should not be too hard I think, I'll add that to the todo list.